lib/Mojolicious/Guides/Testing.pod

```diff
@@ -0,0 +1,754 @@
+                        channel => '+15555555555' })
+    ->status_is(201)
+    ->json_is('/message' => "notification created");
+
+  # Perform GET requests and look at the responses
+  $t->get_ok('/sunrise')
+    ->status_is(200)
+    ->content_like(qr/ am$/);
+
+  $t->get_ok('/sunset')
+    ->status_is(200)
+    ->content_like(qr/ pm$/);
+
+  # Post a URL-encoded form
+  $t->post_ok('/insurance',
+              form => { name   => 'Jimmy',
+                        amount => '3.000.000',
+                        currency => '€' })
+    ->status_is(200);
+
+  # Use Test::More's like() to check the response
+  like $t->tx->res->dom->at('div#thanks')->text, qr(thank you), "thanks";
+
+  done_testing();
+
+In the rest of this document we'll explore these concepts and others related
+to L<Test::Mojo>.
+
+=head1 CONCEPTS
+
+Essentials every L<Mojolicious> developer should know.
+
+=head2 L<Test::Mojo> At a Glance
+
+The L<Test::More> module bundled with Perl includes several primitive test
+assertions, such as C<ok>, C<is>, C<isnt>, C<like>, C<unlike>, C<cmp_ok>,
+etc. An assertion "passes" if its expression returns a true value. The
+assertion method prints "ok" or "not ok" if an assertion passes or fails
+(respectively).
+
+L<Test::Mojo> supplies additional test assertions organized around the web
+application request/response transaction (transport, response headers,
+response bodies, etc.), and WebSocket communications.
+
+One interesting thing of note: the return value of L<Test::Mojo> object
+assertions is always the test object itself, allowing us to "chain" test
+assertion methods. So rather than grouping related test statements like this:
+
+  $t->get_ok('/frogs');
+  $t->status_is(200);
+  $t->content_like(qr/bullfrog/);
+  $t->content_like(qr/hypnotoad/);
+
+method chaining allows us to connect test assertions that belong together:
+
+  $t->get_ok('/frogs')
+    ->status_is(200)
+    ->content_like(qr/bullfrog/)
+    ->content_like(qr/hypnotoad/);
+
+This makes for a much more I<concise> and I<coherent> testing experience:
+concise because we are not repeating the invocant for each test, and coherent
+because assertions that belong to the same request are syntactically bound in
+the same method chain.
+
+Occsionally it makes sense to break up a test to perform more complex
+assertions on a response. L<Test::Mojo> exposes the entire transaction object
+so you can get all the data you need from a response:
+
+  $t->put_ok('/bees', json => { type => 'worker', name => 'Karl' })
+    ->status_is(202)
+    ->json_has('/id');
+
+  # Pull out the id from the response
+  my $newbee = $t->tx->res->json('/id');
+
+  # Make a new request with data from the previous response
+  $t->get_ok("/bees/$newbee")
+    ->status_is(200)
+    ->json_is('/name' => 'Karl');
+
+The L<Test::Mojo> object is I<stateful>. As long as we haven't started a new
+transaction by invoking one of the C<*_ok> methods, the request and response
```

Comment:
I might take it further. Having access to the `$t->tx` means that you can test things that we don't have a convenience method for.

`cmp_ok $t->tx->res->dom->at('epoch'), '>', 0, 'the current epoch value must certainly be before 1970!';`